libcudacxx/include/cuda/std/detail/libcxx/include/__functional/invoke.h

```diff
@@ -481,11 +481,22 @@ struct __invoke_of
 {
 #if defined(__NVCC__) && defined(__CUDACC_EXTENDED_LAMBDA__) && \
    !defined(__CUDA_ARCH__)
-  static_assert(!__nv_is_extended_device_lambda_closure_type(_Fp),
-                "Attempt to use an extended __device__ lambda in a context "
-                "that requires querying its return type in host code. Use a "
-                "named function object, a __host__ __device__ lambda, or "
-                "cuda::proclaim_return_type instead.");
+  #if defined(CUDART_VERSION) && (CUDART_VERSION >= 12030) 
+    static_assert(!__nv_is_extended_device_lambda_closure_type(_Fp) ||
+                   __nv_is_extended_host_device_lambda_closure_type(_Fp) ||
+                   __nv_is_extended_device_lambda_with_preserved_return_type(_Fp),
+                   "Attempt to use an extended __device__ lambda in a context "
+                   "that requires querying its return type in host code. Use a "
+                   "named function object, a __host__ __device__ lambda, or "
+                   "cuda::proclaim_return_type or a extended __device__ lambda "
+                   "with preserved return type instead.");
```

Comment:
The diagnostic name is "preserved return type." However, in my understanding, the C++ feature is typically called a "trailing return type." Should we rename this? I would not know what "preserved" means unless I was familiar with this diagnostic, but I do know what "trailing" means.

```suggestion
                   "with trailing return type instead.");
```

Reply:
Two different things.

Normally, the return type for device lambdas is discarded.

One way of preserving the return type is by using a trailing return type, aka `[] __device__ () -> float`.

There could be other ways of "preserving" the return type in the future, so they named the intrinsic more generically.

Reply:
@Revaj I realized I wasn't very clear here. I agree with @bdice's suggested change. Currently, a trailing return type is the only way to preserve the return type, so I think the diagnostic should be specific.

Reply:
Fixed